opengrok-indexer/src/main/java/org/opengrok/indexer/util/Progress.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2007, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Portions Copyright (c) 2017-2018, Chris Fraire <cfraire@me.com>.
+ */
+
+package org.opengrok.indexer.util;
+
+import org.opengrok.indexer.configuration.RuntimeEnvironment;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class Progress implements AutoCloseable {
+    private Logger logger;
+    private long totalCount;
+    private String prefix;
+
+    private AtomicLong currentCount = new AtomicLong();
+    private Thread loggerThread = null;
+    private AtomicBoolean run = new AtomicBoolean();
+
+    private final Object sync = new Object();
+
+    /**
+     * @param logger logger instance
+     * @param prefix string prefix to identify the operation
+     * @param totalCount total count
+     */
+    public Progress(Logger logger, String prefix, long totalCount) {
+        this.logger = logger;
+        this.prefix = prefix;
+        this.totalCount = totalCount;
+
+        // Assuming printProgress configuration setting cannot be changed on the fly.
+        if (RuntimeEnvironment.getInstance().isPrintProgress()) {
+            // spawn a logger thread.
+            run.set(true);
+            loggerThread = new Thread(this::logLoop,
+                    "progress-thread-" + prefix.replaceAll(" ", "_"));
+            loggerThread.start();
+        }
+    }
+
+    /**
+     * Increment counter. The actual logging will be done eventually.
+     */
+    public void increment() {
+        this.currentCount.incrementAndGet();
+
+        if (loggerThread != null) {
+            // nag the thread.
+            synchronized (sync) {
+                sync.notify();
+            }
+        }
+    }
+
+    private void logLoop() {
+        while (run.get()) {
+            long currentCount = this.currentCount.get();
+            Level currentLevel;
+
+            if (currentCount <= 1 || currentCount >= totalCount ||
+                    currentCount % 100 == 0) {
+                currentLevel = Level.INFO;
+            } else if (currentCount % 50 == 0) {
+                currentLevel = Level.FINE;
+            } else if (currentCount % 10 == 0) {
+                currentLevel = Level.FINER;
+            } else {
+                currentLevel = Level.FINEST;
+            }
+
+            if (logger.isLoggable(currentLevel)) {
+                logger.log(currentLevel, "Progress: {0} ({1}%) for {2}",
+                        new Object[]{currentCount, currentCount * 100.0f /
+                                totalCount, prefix});
+            }
+
+            if (currentCount >= totalCount) {
+                return;
+            }
+
+            // wait for event
+            try {
+                synchronized (sync) {
+                    if (!run.get()) {
```

Comment:
Do you think you might accidentally forgo logging the final entry by returning here?

Reply:
It's actually deadlock prevention. If the thread calling close() notifies the loop thread while it is still executing the loop body (before the wait() call) they will wait for each other indefinitely.